Compilation error when running "game.sh" at 4%

I have run dependencies.sh from https://github.com/RigsOfRods/ror-linux-buildscripts and am now running game.sh, but it exits with the following error:

`[  4%] Linking CXX executable ../../bin/RoRConfig
/home/victoria/rigs-of-rods-source/rigs-of-rods/source/angelscript_addons/scriptstdstring/scriptstdstring.cpp: In function ‘void AngelScript::RegisterStdString_Native(AngelScript::asIScriptEngine*)’:
/home/victoria/rigs-of-rods-source/rigs-of-rods/source/angelscript_addons/scriptstdstring/scriptstdstring.cpp:744:93: error: no matching function for call to ‘AngelScript::asIScriptEngine::RegisterStringFactory(const char [15], AngelScript::asSFuncPtr, AngelScript::asECallConvTypes)’
  r = engine->RegisterStringFactory("const string &", asFUNCTION(StringFactory), asCALL_CDECL); assert( r >= 0 );`
       

How can I make the compilation work? It seems the function "engine->RegisterStringFactory("const string &", asFUNCTION(StringFactory), asCALL_CDECL)" does not exist, even though dependencies are installed.

I am using Debian 9, with the latest version of CMake and the latest version of game.sh

Thanks.

--- a/source/angelscript_addons/scriptany/scriptany.h
+++ b/source/angelscript_addons/scriptany/scriptany.h
@@ -9,7 +9,7 @@
 
 BEGIN_AS_NAMESPACE
 
-class CScriptAny
+class CScriptAny 
 {
 public:
 	// Constructors

--- a/source/angelscript_addons/scriptarray/scriptarray.cpp
+++ b/source/angelscript_addons/scriptarray/scriptarray.cpp
@@ -316,6 +316,10 @@ static void RegisterScriptArray_Native(asIScriptEngine *engine)
 	r = engine->RegisterObjectMethod("array<T>", "bool opEquals(const array<T>&in) const", asMETHOD(CScriptArray, operator==), asCALL_THISCALL); assert( r >= 0 );
 	r = engine->RegisterObjectMethod("array<T>", "bool isEmpty() const", asMETHOD(CScriptArray, IsEmpty), asCALL_THISCALL); assert( r >= 0 );
 
+	// Sort with callback for comparison
+	r = engine->RegisterFuncdef("bool array<T>::less(const T&in a, const T&in b)");
+	r = engine->RegisterObjectMethod("array<T>", "void sort(const less &in, uint startAt = 0, uint count = uint(-1))", asMETHODPR(CScriptArray, Sort, (asIScriptFunction*, asUINT, asUINT), void), asCALL_THISCALL); assert(r >= 0);
+
 	// Register virtual properties
 	r = engine->RegisterObjectMethod("array<T>", "uint get_length() const", asMETHOD(CScriptArray, GetSize), asCALL_THISCALL); assert( r >= 0 );
 	r = engine->RegisterObjectMethod("array<T>", "void set_length(uint)", asMETHODPR(CScriptArray, Resize, (asUINT), void), asCALL_THISCALL); assert( r >= 0 );
@@ -665,7 +669,7 @@ void CScriptArray::RemoveRange(asUINT start, asUINT count)
 	// Compact the elements
 	// As objects in arrays of objects are not stored inline, it is safe to use memmove here
 	// since we're just copying the pointers to objects and not the actual objects.
-	memmove(buffer->data + start*elementSize, buffer->data + (start + count)*elementSize, count*elementSize);
+	memmove(buffer->data + start*elementSize, buffer->data + (start + count)*elementSize, (buffer->numElements - start - count)*elementSize);
 	buffer->numElements -= count;
 }
 
@@ -1483,10 +1487,7 @@ void CScriptArray::Sort(asUINT startAt, asUINT count, bool asc)
 		}
 		if( cmpContext == 0 )
 		{
-			// TODO: Ideally this context would be retrieved from a pool, so we don't have to
-			//       create a new one everytime. We could keep a context with the array object
-			//       but that would consume a lot of resources as each context is quite heavy.
-			cmpContext = objType->GetEngine()->CreateContext();
+			cmpContext = objType->GetEngine()->RequestContext();
 		}
 	}
 
@@ -1516,7 +1517,88 @@ void CScriptArray::Sort(asUINT startAt, asUINT count, bool asc)
 				cmpContext->Abort();
 		}
 		else
-			cmpContext->Release();
+			objType->GetEngine()->ReturnContext(cmpContext);
+	}
+}
+
+// Sort with script callback for comparing elements
+void CScriptArray::Sort(asIScriptFunction *func, asUINT startAt, asUINT count)
+{
+	// No need to sort
+	if (count < 2)
+		return;
+
+	// Check if we could access invalid item while sorting
+	asUINT start = startAt;
+	asUINT end = asQWORD(startAt) + asQWORD(count) >= (asQWORD(1)<<32) ? 0xFFFFFFFF : startAt + count;
+	if (end > buffer->numElements)
+		end = buffer->numElements;
+
+	if (start >= buffer->numElements)
+	{
+		asIScriptContext *ctx = asGetActiveContext();
+
+		// Throw an exception
+		if (ctx)
+			ctx->SetException("Index out of bounds");
+
+		return;
+	}
+
+	asBYTE tmp[16];
+	asIScriptContext *cmpContext = 0;
+	bool isNested = false;
+
+	// Try to reuse the active context
+	cmpContext = asGetActiveContext();
+	if (cmpContext)
+	{
+		if (cmpContext->GetEngine() == objType->GetEngine() && cmpContext->PushState() >= 0)
+			isNested = true;
+		else
+			cmpContext = 0;
+	}
+	if (cmpContext == 0)
+		cmpContext = objType->GetEngine()->RequestContext();
+
+	// Insertion sort
+	for (asUINT i = start + 1; i < end; i++)
+	{
+		Copy(tmp, GetArrayItemPointer(i));
+
+		asUINT j = i - 1;
+
+		while (j != 0xFFFFFFFF && j >= start )
+		{
+			cmpContext->Prepare(func);
+			cmpContext->SetArgAddress(0, GetDataPointer(tmp));
+			cmpContext->SetArgAddress(1, At(j));
+			int r = cmpContext->Execute();
+			if (r != asEXECUTION_FINISHED)
+				break;
+			if (*(bool*)(cmpContext->GetAddressOfReturnValue()))
+			{
+				Copy(GetArrayItemPointer(j + 1), GetArrayItemPointer(j));
+				j--;
+			}
+			else
+				break;
+		}
+
+		Copy(GetArrayItemPointer(j + 1), tmp);
+	}
+
+	if (cmpContext)
+	{
+		if (isNested)
+		{
+			asEContextState state = cmpContext->GetState();
+			cmpContext->PopState();
+			if (state == asEXECUTION_ABORTED)
+				cmpContext->Abort();
+		}
+		else
+			objType->GetEngine()->ReturnContext(cmpContext);
 	}
 }
 
@@ -1955,6 +2037,15 @@ static void ScriptArraySortDesc2_Generic(asIScriptGeneric *gen)
 	self->SortDesc(index, count);
 }
 
+static void ScriptArraySortCallback_Generic(asIScriptGeneric *gen)
+{
+	asIScriptFunction *callback = (asIScriptFunction*)gen->GetArgAddress(0);
+	asUINT startAt = gen->GetArgDWord(1);
+	asUINT count = gen->GetArgDWord(2);
+	CScriptArray *self = (CScriptArray*)gen->GetObject();
+	self->Sort(callback, startAt, count);
+}
+
 static void ScriptArrayAddRef_Generic(asIScriptGeneric *gen)
 {
 	CScriptArray *self = (CScriptArray*)gen->GetObject();
@@ -2039,6 +2130,8 @@ static void RegisterScriptArray_Generic(asIScriptEngine *engine)
 	r = engine->RegisterObjectMethod("array<T>", "int findByRef(uint startAt, const T&in if_handle_then_const value) const", asFUNCTION(ScriptArrayFindByRef2_Generic), asCALL_GENERIC); assert( r >= 0 );
 	r = engine->RegisterObjectMethod("array<T>", "bool opEquals(const array<T>&in) const", asFUNCTION(ScriptArrayEquals_Generic), asCALL_GENERIC); assert( r >= 0 );
 	r = engine->RegisterObjectMethod("array<T>", "bool isEmpty() const", asFUNCTION(ScriptArrayIsEmpty_Generic), asCALL_GENERIC); assert( r >= 0 );
+	r = engine->RegisterFuncdef("bool array<T>::less(const T&in a, const T&in b)");
+	r = engine->RegisterObjectMethod("array<T>", "void sort(const less &in, uint startAt = 0, uint count = uint(-1))", asFUNCTION(ScriptArraySortCallback_Generic), asCALL_GENERIC); assert(r >= 0);
 	r = engine->RegisterObjectMethod("array<T>", "uint get_length() const", asFUNCTION(ScriptArrayLength_Generic), asCALL_GENERIC); assert( r >= 0 );
 	r = engine->RegisterObjectMethod("array<T>", "void set_length(uint)", asFUNCTION(ScriptArrayResize_Generic), asCALL_GENERIC); assert( r >= 0 );
 	r = engine->RegisterObjectBehaviour("array<T>", asBEHAVE_GETREFCOUNT, "int f()", asFUNCTION(ScriptArrayGetRefCount_Generic), asCALL_GENERIC); assert( r >= 0 );

--- a/source/angelscript_addons/scriptarray/scriptarray.h
+++ b/source/angelscript_addons/scriptarray/scriptarray.h
@@ -83,6 +83,7 @@ public:
 	void SortAsc(asUINT startAt, asUINT count);
 	void SortDesc(asUINT startAt, asUINT count);
 	void Sort(asUINT startAt, asUINT count, bool asc);
+	void Sort(asIScriptFunction *less, asUINT startAt, asUINT count);
 	void Reverse();
 	int  Find(void *value) const;
 	int  Find(asUINT startAt, void *value) const;

--- a/source/angelscript_addons/scriptbuilder/scriptbuilder.cpp
+++ b/source/angelscript_addons/scriptbuilder/scriptbuilder.cpp
@@ -283,7 +283,7 @@ int CScriptBuilder::ProcessScriptSection(const char *script, unsigned int length
 
 #if AS_PROCESS_METADATA == 1
 	// Preallocate memory
-	string metadata, declaration;
+	string metadata, name, declaration;
 	metadata.reserve(500);
 	declaration.reserve(100);
 #endif
@@ -320,7 +320,7 @@ int CScriptBuilder::ProcessScriptSection(const char *script, unsigned int length
 			{
 				currentClass = modifiedScript.substr(pos,len);
 
-				// Search until first { is encountered
+				// Search until first { or ; is encountered
 				while( pos < modifiedScript.length() )
 				{
 					engine->ParseToken(&modifiedScript[pos], modifiedScript.size() - pos, &len);
@@ -328,6 +328,13 @@ int CScriptBuilder::ProcessScriptSection(const char *script, unsigned int length
 					// If start of class section encountered stop
 					if( modifiedScript[pos] == '{' )
 					{
+						pos += len;
+						break;
+					}
+					else if (modifiedScript[pos] == ';')
+					{
+						// The class declaration has ended and there are no children
+						currentClass = "";
 						pos += len;
 						break;
 					}
@@ -405,12 +412,12 @@ int CScriptBuilder::ProcessScriptSection(const char *script, unsigned int length
 
 			// Determine what this metadata is for
 			int type;
-			ExtractDeclaration(pos, declaration, type);
+			ExtractDeclaration(pos, name, declaration, type);
 
 			// Store away the declaration in a map for lookup after the build has completed
 			if( type > 0 )
 			{
-				SMetadataDecl decl(metadata, declaration, type, currentClass, currentNamespace);
+				SMetadataDecl decl(metadata, name, declaration, type, currentClass, currentNamespace);
 				foundDeclarations.push_back(decl);
 			}
 		}
@@ -521,7 +528,7 @@ int CScriptBuilder::Build()
 	{
 		SMetadataDecl *decl = &foundDeclarations[n];
 		module->SetDefaultNamespace(decl->nameSpace.c_str());
-		if( decl->type == 1 )
+		if( decl->type == MDT_TYPE )
 		{
 			// Find the type id
 			int typeId = module->GetTypeIdByDecl(decl->declaration.c_str());
@@ -529,7 +536,7 @@ int CScriptBuilder::Build()
 			if( typeId >= 0 )
 				typeMetadataMap.insert(map<int, string>::value_type(typeId, decl->metadata));
 		}
-		else if( decl->type == 2 )
+		else if( decl->type == MDT_FUNC )
 		{
 			if( decl->parentClass == "" )
 			{
@@ -558,7 +565,7 @@ int CScriptBuilder::Build()
 					it->second.funcMetadataMap.insert(map<int, string>::value_type(func->GetId(), decl->metadata));
 			}
 		}
-		else if( decl->type == 4 )
+		else if( decl->type == MDT_VIRTPROP )
 		{
 			if( decl->parentClass == "" )
 			{
@@ -589,10 +596,9 @@ int CScriptBuilder::Build()
 				func = type->GetMethodByName(("set_" + decl->declaration).c_str());
 				if( func )
 					it->second.funcMetadataMap.insert(map<int, string>::value_type(func->GetId(), decl->metadata));
-
 			}
 		}
-		else if( decl->type == 3 )
+		else if( decl->type == MDT_VAR )
 		{
 			if( decl->parentClass == "" )
 			{
@@ -634,6 +640,65 @@ int CScriptBuilder::Build()
 				// If found, add it
 				assert( idx >= 0 );
 				if( idx >= 0 ) it->second.varMetadataMap.insert(map<int, string>::value_type(idx, decl->metadata));
+			}
+		}
+		else if (decl->type == MDT_FUNC_OR_VAR)
+		{
+			if (decl->parentClass == "")
+			{
+				// Find the global variable index
+				int varIdx = module->GetGlobalVarIndexByName(decl->name.c_str());
+				if (varIdx >= 0)
+					varMetadataMap.insert(map<int, string>::value_type(varIdx, decl->metadata));
+				else
+				{
+					asIScriptFunction *func = module->GetFunctionByDecl(decl->declaration.c_str());
+					assert(func);
+					if (func)
+						funcMetadataMap.insert(map<int, string>::value_type(func->GetId(), decl->metadata));
+				}
+			}
+			else
+			{
+				int typeId = module->GetTypeIdByDecl(decl->parentClass.c_str());
+				assert(typeId > 0);
+
+				// Add the classes if needed
+				map<int, SClassMetadata>::iterator it = classMetadataMap.find(typeId);
+				if (it == classMetadataMap.end())
+				{
+					classMetadataMap.insert(map<int, SClassMetadata>::value_type(typeId, SClassMetadata(decl->parentClass)));
+					it = classMetadataMap.find(typeId);
+				}
+
+				// Add the variable to class
+				asITypeInfo *objectType = engine->GetTypeInfoById(typeId);
+				int idx = -1;
+
+				// Search through all properties to get proper declaration
+				for (asUINT i = 0; i < (asUINT)objectType->GetPropertyCount(); ++i)
+				{
+					const char *name;
+					objectType->GetProperty(i, &name);
+					if (decl->name == name)
+					{
+						idx = i;
+						break;
+					}
+				}
+
+				// If found, add it
+				if (idx >= 0) 
+					it->second.varMetadataMap.insert(map<int, string>::value_type(idx, decl->metadata));
+				else
+				{
+					// Look for the matching method instead
+					asITypeInfo *type = engine->GetTypeInfoById(typeId);
+					asIScriptFunction *func = type->GetMethodByDecl(decl->declaration.c_str());
+					assert(func);
+					if (func)
+						it->second.funcMetadataMap.insert(map<int, string>::value_type(func->GetId(), decl->metadata));
+				}
 			}
 		}
 	}
@@ -773,7 +838,7 @@ int CScriptBuilder::ExtractMetadataString(int pos, string &metadata)
 	return pos;
 }
 
-int CScriptBuilder::ExtractDeclaration(int pos, string &declaration, int &type)
+int CScriptBuilder::ExtractDeclaration(int pos, string &name, string &declaration, int &type)
 {
 	declaration = "";
 	type = 0;
@@ -784,18 +849,19 @@ int CScriptBuilder::ExtractDeclaration(int pos, string &declaration, int &type)
 	asUINT len = 0;
 	asETokenClass t = asTC_WHITESPACE;
 
-	// Skip white spaces and comments
+	// Skip white spaces, comments, and leading 'shared', 'external', 'final', 'abstract'
 	do
 	{
 		pos += len;
 		t = engine->ParseToken(&modifiedScript[pos], modifiedScript.size() - pos, &len);
-	} while ( t == asTC_WHITESPACE || t == asTC_COMMENT );
+		token.assign(&modifiedScript[pos], len);
+	} while ( t == asTC_WHITESPACE || t == asTC_COMMENT || token == "shared" || token == "external" || token == "final" || token == "abstract" );
 
 	// We're expecting, either a class, interface, function, or variable declaration
 	if( t == asTC_KEYWORD || t == asTC_IDENTIFIER )
 	{
 		token.assign(&modifiedScript[pos], len);
-		if( token == "interface" || token == "class" )
+		if( token == "interface" || token == "class" || token == "enum" )
 		{
 			// Skip white spaces and comments
 			do
@@ -806,7 +872,7 @@ int CScriptBuilder::ExtractDeclaration(int pos, string &declaration, int &type)
 
 			if( t == asTC_IDENTIFIER )
 			{
-				type = 1;
+				type = MDT_TYPE;
 				declaration.assign(&modifiedScript[pos], len);
 				pos += len;
 				return pos;
@@ -822,7 +888,6 @@ int CScriptBuilder::ExtractDeclaration(int pos, string &declaration, int &type)
 			bool hasParenthesis = false;
 			declaration.append(&modifiedScript[pos], len);
 			pos += len;
-			string name;
 			for(; pos < (int)modifiedScript.size();)
 			{
 				t = engine->ParseToken(&modifiedScript[pos], modifiedScript.size() - pos, &len);
@@ -834,21 +899,29 @@ int CScriptBuilder::ExtractDeclaration(int pos, string &declaration, int &type)
 						if( hasParenthesis )
 						{
 							// We've found the end of a function signature
-							type = 2;
+							type = MDT_FUNC;
 						}
 						else
 						{
 							// We've found a virtual property. Just keep the name
 							declaration = name;
-							type = 4;
+							type = MDT_VIRTPROP;
 						}
 						return pos;
 					}
 					if( (token == "=" && !hasParenthesis) || token == ";" )
 					{
-						// Substitute the declaration with just the name
-						declaration = name;
-						type = 3;
+						if (hasParenthesis)
+						{
+							// The declaration is ambigous. It can be a variable with initialization, or a function prototype
+							type = MDT_FUNC_OR_VAR; 
+						}
+						else
+						{
+							// Substitute the declaration with just the name
+							declaration = name;
+							type = MDT_VAR;
+						}
 						return pos;
 					}
 					else if( token == "(" )

--- a/source/angelscript_addons/scriptbuilder/scriptbuilder.h
+++ b/source/angelscript_addons/scriptbuilder/scriptbuilder.h
@@ -1,6 +1,7 @@
 #ifndef SCRIPTBUILDER_H
 #define SCRIPTBUILDER_H
 
+
 // ###########################
 // Rigs of Rods - notes
 //
@@ -95,7 +96,7 @@ public:
 	std::string  GetSectionName(unsigned int idx) const;
 
 #if AS_PROCESS_METADATA == 1
-	// Get metadata declared for class types and interfaces
+	// Get metadata declared for classes, interfaces, and enums
 	const char *GetMetadataStringForType(int typeId);
 
 	// Get metadata declared for functions
@@ -107,7 +108,7 @@ public:
 	// Get metadata declared for class variables
 	const char *GetMetadataStringForTypeProperty(int typeId, int varIdx);
 
-	// Get metadata declared for class functions
+	// Get metadata declared for class methods
 	const char *GetMetadataStringForTypeMethod(int typeId, asIScriptFunction *method);
 #endif
 
@@ -115,7 +116,7 @@ protected:
 	void ClearAll();
 	int  Build();
 	int  ProcessScriptSection(const char *script, unsigned int length, const char *sectionname, int lineOffset);
-	virtual int  LoadScriptSection(const char *filename); // ## Rigs of Rods modification
+	virtual int  LoadScriptSection(const char *filename);
 	bool IncludeIfNotAlreadyIncluded(const char *filename);
 
 	int  SkipStatement(int pos);
@@ -132,13 +133,23 @@ protected:
 
 #if AS_PROCESS_METADATA == 1
 	int  ExtractMetadataString(int pos, std::string &outMetadata);
-	int  ExtractDeclaration(int pos, std::string &outDeclaration, int &outType);
+	int  ExtractDeclaration(int pos, std::string &outName, std::string &outDeclaration, int &outType);
+
+	enum METADATATYPE
+	{
+		MDT_TYPE = 1,
+		MDT_FUNC = 2,
+		MDT_VAR = 3,
+		MDT_VIRTPROP = 4,
+		MDT_FUNC_OR_VAR = 5
+	};
 
 	// Temporary structure for storing metadata and declaration
 	struct SMetadataDecl
 	{
-		SMetadataDecl(std::string m, std::string d, int t, std::string c, std::string ns) : metadata(m), declaration(d), type(t), parentClass(c), nameSpace(ns) {}
+		SMetadataDecl(std::string m, std::string n, std::string d, int t, std::string c, std::string ns) : metadata(m), name(n), declaration(d), type(t), parentClass(c), nameSpace(ns) {}
 		std::string metadata;
+		std::string name;
 		std::string declaration;
 		int         type;
 		std::string parentClass;

--- a/source/angelscript_addons/scriptdictionary/scriptdictionary.cpp
+++ b/source/angelscript_addons/scriptdictionary/scriptdictionary.cpp
@@ -695,7 +695,7 @@ bool CScriptDictValue::Get(asIScriptEngine *engine, void *value, int typeId) con
 		bool isCompatible = false;
 
 		// Allow a handle to be value assigned if the wanted type is not a handle
-		if( (m_typeId & ~asTYPEID_OBJHANDLE) == typeId && m_valueObj != 0 )
+		if( (m_typeId & ~(asTYPEID_OBJHANDLE | asTYPEID_HANDLETOCONST) ) == typeId && m_valueObj != 0 )
 			isCompatible = true;
 
 		// Copy the object into the given reference

--- a/source/angelscript_addons/scripthelper/scripthelper.cpp
+++ b/source/angelscript_addons/scripthelper/scripthelper.cpp
@@ -245,7 +245,7 @@ int WriteConfigToStream(asIScriptEngine *engine, ostream &strm)
 
 	asDWORD currAccessMask = 0;
 	string currNamespace = "";
-	engine->SetDefaultNamespace(currNamespace.c_str());
+	engine->SetDefaultNamespace("");
 
 	// Export the engine version, just for info
 	strm << "// AngelScript " << asGetLibraryVersion() << "\n";
@@ -449,7 +449,13 @@ int WriteConfigToStream(asIScriptEngine *engine, ostream &strm)
 						strm << "access " << hex << (unsigned int)(accessMask) << dec << "\n";
 						currAccessMask = accessMask;
 					}
-					strm << "objprop \"" << typeDecl.c_str() << "\" \"" << type->GetPropertyDeclaration(m) << "\"\n";
+					strm << "objprop \"" << typeDecl.c_str() << "\" \"" << type->GetPropertyDeclaration(m) << "\"";
+
+					// Save information about composite properties
+					int compositeOffset;
+					bool isCompositeIndirect;
+					type->GetProperty(m, 0, 0, 0, 0, 0, 0, 0, &compositeOffset, &isCompositeIndirect);
+					strm << " " << compositeOffset << " " << (isCompositeIndirect ? "1" : "0") << "\n";
 				}
 			}
 		}
@@ -524,10 +530,17 @@ int WriteConfigToStream(asIScriptEngine *engine, ostream &strm)
 		strm << "prop \"" << (isConst ? "const " : "") << engine->GetTypeDeclaration(typeId) << " " << name << "\"\n";
 	}
 
-	engine->SetDefaultNamespace("");
-
 	// Write string factory
 	strm << "\n// String factory\n";
+
+	// Reset the namespace for the string factory and default array type
+	if ("" != currNamespace)
+	{
+		strm << "namespace \"\"\n";
+		currNamespace = "";
+		engine->SetDefaultNamespace("");
+	}
+
 	asDWORD flags = 0;
 	int typeId = engine->GetStringFactoryReturnTypeId(&flags);
 	if( typeId > 0 )
@@ -545,7 +558,7 @@ int WriteConfigToStream(asIScriptEngine *engine, ostream &strm)
 	return 0;
 }
 
-int ConfigEngineFromStream(asIScriptEngine *engine, istream &strm, const char *configFile)
+int ConfigEngineFromStream(asIScriptEngine *engine, istream &strm, const char *configFile, asIStringFactory *stringFactory)
 {
 	int r;
 
@@ -725,11 +738,13 @@ int ConfigEngineFromStream(asIScriptEngine *engine, istream &strm, const char *c
 		}
 		else if( token == "objprop" )
 		{
-			string name, decl;
+			string name, decl, compositeOffset, isCompositeIndirect;
 			in::GetToken(engine, name, config, pos);
 			name = name.substr(1, name.length() - 2);
 			in::GetToken(engine, decl, config, pos);
 			decl = decl.substr(1, decl.length() - 2);
+			in::GetToken(engine, compositeOffset, config, pos);
+			in::GetToken(engine, isCompositeIndirect, config, pos);
 
 			asITypeInfo *type = engine->GetTypeInfoById(engine->GetTypeIdByDecl(name.c_str()));
 			if( type == 0 )
@@ -740,7 +755,7 @@ int ConfigEngineFromStream(asIScriptEngine *engine, istream &strm, const char *c
 
 			// All properties must have different offsets in order to make them
 			// distinct, so we simply register them with an incremental offset
-			r = engine->RegisterObjectProperty(name.c_str(), decl.c_str(), type->GetPropertyCount());
+			r = engine->RegisterObjectProperty(name.c_str(), decl.c_str(), type->GetPropertyCount(), compositeOffset != "0" ? type->GetPropertyCount() : 0, isCompositeIndirect != "0");
 			if( r < 0 )
 			{
 				engine->WriteMessage(configFile, in::GetLineNumber(config, pos), 0, asMSGTYPE_ERROR, "Failed to register object property");
@@ -810,11 +825,19 @@ int ConfigEngineFromStream(asIScriptEngine *engine, istream &strm, const char *c
 			in::GetToken(engine, type, config, pos);
 			type = type.substr(1, type.length() - 2);
 
-			r = engine->RegisterStringFactory(type.c_str(), asFUNCTION(0), asCALL_GENERIC);
-			if( r < 0 )
+			if (stringFactory == 0)
 			{
-				engine->WriteMessage(configFile, in::GetLineNumber(config, pos), 0, asMSGTYPE_ERROR, "Failed to register string factory");
+				engine->WriteMessage(configFile, in::GetLineNumber(config, pos), 0, asMSGTYPE_WARNING, "Cannot register string factory without the actual implementation");
 				return -1;
+			}
+			else
+			{
+				r = engine->RegisterStringFactory(type.c_str(), stringFactory);
+				if (r < 0)
+				{
+					engine->WriteMessage(configFile, in::GetLineNumber(config, pos), 0, asMSGTYPE_ERROR, "Failed to register string factory");
+					return -1;
+				}
 			}
 		}
 		else if( token == "defarray" )

--- a/source/angelscript_addons/scripthelper/scripthelper.h
+++ b/source/angelscript_addons/scripthelper/scripthelper.h
@@ -1,7 +1,6 @@
 #ifndef SCRIPTHELPER_H
 #define SCRIPTHELPER_H
 
-
 // ##################################
 // RoR notes, last updated 2017-08-30
 //
@@ -39,13 +38,13 @@ int ExecuteString(asIScriptEngine *engine, const char *code, void *ret, int retT
 // The format is compatible with the offline compiler in /sdk/samples/asbuild/.
 int WriteConfigToFile(asIScriptEngine *engine, const char *filename);
 
-// Write the registered application interface to a text stream. 
-int WriteConfigToStream(asIScriptEngine *engine, std::ostream &strm); 
+// Write the registered application interface to a text stream.
+int WriteConfigToStream(asIScriptEngine *engine, std::ostream &strm);
 
-// Loads an interface from a text stream and configures the engine with it. This will not 
+// Loads an interface from a text stream and configures the engine with it. This will not
 // set the correct function pointers, so it is not possible to use this engine to execute
 // scripts, but it can be used to compile scripts and save the byte code.
-int ConfigEngineFromStream(asIScriptEngine *engine, std::istream &strm, const char *nameOfStream = "config");
+int ConfigEngineFromStream(asIScriptEngine *engine, std::istream &strm, const char *nameOfStream = "config", asIStringFactory *stringFactory = 0);
 
 // Format the details of the script exception into a human readable text
 std::string GetExceptionInfo(asIScriptContext *ctx, bool showStack = false);

--- a/source/angelscript_addons/scriptmath/scriptmathcomplex.cpp
+++ b/source/angelscript_addons/scriptmath/scriptmathcomplex.cpp
@@ -1,0 +1,222 @@
+#include <assert.h>
+#include <string.h> // strstr
+#include <new> // new()
+#include <math.h>
+#include "scriptmathcomplex.h"
+
+#ifdef __BORLANDC__
+// C++Builder doesn't define a non-standard "sqrtf" function but rather an overload of "sqrt"
+// for float arguments.
+inline float sqrtf (float x) { return sqrt (x); }
+#endif
+
+BEGIN_AS_NAMESPACE
+
+Complex::Complex()
+{
+	r = 0;
+	i = 0;
+}
+
+Complex::Complex(const Complex &other)
+{
+	r = other.r;
+	i = other.i;
+}
+
+Complex::Complex(float _r, float _i)
+{
+	r = _r;
+	i = _i;
+}
+
+bool Complex::operator==(const Complex &o) const
+{
+	return (r == o.r) && (i == o.i);
+}
+
+bool Complex::operator!=(const Complex &o) const
+{
+	return !(*this == o);
+}
+
+Complex &Complex::operator=(const Complex &other)
+{
+	r = other.r;
+	i = other.i;
+	return *this;
+}
+
+Complex &Complex::operator+=(const Complex &other)
+{
+	r += other.r;
+	i += other.i;
+	return *this;
+}
+
+Complex &Complex::operator-=(const Complex &other)
+{
+	r -= other.r;
+	i -= other.i;
+	return *this;
+}
+
+Complex &Complex::operator*=(const Complex &other)
+{
+	*this = *this * other;
+	return *this;
+}
+
+Complex &Complex::operator/=(const Complex &other)
+{
+	*this = *this / other;
+	return *this;
+}
+
+float Complex::squaredLength() const
+{
+	return r*r + i*i;
+}
+
+float Complex::length() const
+{
+	return sqrtf(squaredLength());
+}
+
+Complex Complex::operator+(const Complex &other) const
+{
+	return Complex(r + other.r, i + other.i);
+}
+
+Complex Complex::operator-(const Complex &other) const
+{
+	return Complex(r - other.r, i + other.i);
+}
+
+Complex Complex::operator*(const Complex &other) const
+{
+	return Complex(r*other.r - i*other.i, r*other.i + i*other.r);
+}
+
+Complex Complex::operator/(const Complex &other) const
+{
+	float squaredLen = other.squaredLength();
+	if( squaredLen == 0 ) return Complex(0,0);
+
+	return Complex((r*other.r + i*other.i)/squaredLen, (i*other.r - r*other.i)/squaredLen);
+}
+
+//-----------------------
+// Swizzle operators
+//-----------------------
+
+Complex Complex::get_ri() const
+{
+	return *this;
+}
+Complex Complex::get_ir() const
+{
+	return Complex(r,i);
+}
+void Complex::set_ri(const Complex &o)
+{
+	*this = o;
+}
+void Complex::set_ir(const Complex &o)
+{
+	r = o.i;
+	i = o.r;
+}
+
+//-----------------------
+// AngelScript functions
+//-----------------------
+
+static void ComplexDefaultConstructor(Complex *self)
+{
+	new(self) Complex();
+}
+
+static void ComplexCopyConstructor(const Complex &other, Complex *self)
+{
+	new(self) Complex(other);
+}
+
+static void ComplexConvConstructor(float r, Complex *self)
+{
+	new(self) Complex(r);
+}
+
+static void ComplexInitConstructor(float r, float i, Complex *self)
+{
+	new(self) Complex(r,i);
+}
+
+static void ComplexListConstructor(float *list, Complex *self)
+{
+	new(self) Complex(list[0], list[1]);
+}
+
+//--------------------------------
+// Registration
+//-------------------------------------
+
+static void RegisterScriptMathComplex_Native(asIScriptEngine *engine)
+{
+	int r;
+
+	// Register the type
+#if AS_CAN_USE_CPP11
+	// With C++11 it is possible to use asGetTypeTraits to determine the correct flags to represent the C++ class, except for the asOBJ_APP_CLASS_ALLFLOATS
+	r = engine->RegisterObjectType("complex", sizeof(Complex), asOBJ_VALUE | asOBJ_POD | asGetTypeTraits<Complex>() | asOBJ_APP_CLASS_ALLFLOATS); assert( r >= 0 );
+#else
+	r = engine->RegisterObjectType("complex", sizeof(Complex), asOBJ_VALUE | asOBJ_POD | asOBJ_APP_CLASS_CAK | asOBJ_APP_CLASS_ALLFLOATS); assert( r >= 0 );
+#endif
+
+	// Register the object properties
+	r = engine->RegisterObjectProperty("complex", "float r", asOFFSET(Complex, r)); assert( r >= 0 );
+	r = engine->RegisterObjectProperty("complex", "float i", asOFFSET(Complex, i)); assert( r >= 0 );
+
+	// Register the constructors
+	r = engine->RegisterObjectBehaviour("complex", asBEHAVE_CONSTRUCT,      "void f()",                             asFUNCTION(ComplexDefaultConstructor), asCALL_CDECL_OBJLAST); assert( r >= 0 );
+	r = engine->RegisterObjectBehaviour("complex", asBEHAVE_CONSTRUCT,      "void f(const complex &in)",            asFUNCTION(ComplexCopyConstructor), asCALL_CDECL_OBJLAST); assert( r >= 0 );
+	r = engine->RegisterObjectBehaviour("complex", asBEHAVE_CONSTRUCT,      "void f(float)",                        asFUNCTION(ComplexConvConstructor), asCALL_CDECL_OBJLAST); assert( r >= 0 );
+	r = engine->RegisterObjectBehaviour("complex", asBEHAVE_CONSTRUCT,      "void f(float, float)",                 asFUNCTION(ComplexInitConstructor), asCALL_CDECL_OBJLAST); assert( r >= 0 );
+	r = engine->RegisterObjectBehaviour("complex", asBEHAVE_LIST_CONSTRUCT, "void f(const int &in) {float, float}", asFUNCTION(ComplexListConstructor), asCALL_CDECL_OBJLAST); assert( r >= 0 );
+
+	// Register the operator overloads
+	r = engine->RegisterObjectMethod("complex", "complex &opAddAssign(const complex &in)", asMETHODPR(Complex, operator+=, (const Complex &), Complex&), asCALL_THISCALL); assert( r >= 0 );
+	r = engine->RegisterObjectMethod("complex", "complex &opSubAssign(const complex &in)", asMETHODPR(Complex, operator-=, (const Complex &), Complex&), asCALL_THISCALL); assert( r >= 0 );
+	r = engine->RegisterObjectMethod("complex", "complex &opMulAssign(const complex &in)", asMETHODPR(Complex, operator*=, (const Complex &), Complex&), asCALL_THISCALL); assert( r >= 0 );
+	r = engine->RegisterObjectMethod("complex", "complex &opDivAssign(const complex &in)", asMETHODPR(Complex, operator/=, (const Complex &), Complex&), asCALL_THISCALL); assert( r >= 0 );
+	r = engine->RegisterObjectMethod("complex", "bool opEquals(const complex &in) const", asMETHODPR(Complex, operator==, (const Complex &) const, bool), asCALL_THISCALL); assert( r >= 0 );
+	r = engine->RegisterObjectMethod("complex", "complex opAdd(const complex &in) const", asMETHODPR(Complex, operator+, (const Complex &) const, Complex), asCALL_THISCALL); assert( r >= 0 );
+	r = engine->RegisterObjectMethod("complex", "complex opSub(const complex &in) const", asMETHODPR(Complex, operator-, (const Complex &) const, Complex), asCALL_THISCALL); assert( r >= 0 );
+	r = engine->RegisterObjectMethod("complex", "complex opMul(const complex &in) const", asMETHODPR(Complex, operator*, (const Complex &) const, Complex), asCALL_THISCALL); assert( r >= 0 );
+	r = engine->RegisterObjectMethod("complex", "complex opDiv(const complex &in) const", asMETHODPR(Complex, operator/, (const Complex &) const, Complex), asCALL_THISCALL); assert( r >= 0 );
+
+	// Register the object methods
+	r = engine->RegisterObjectMethod("complex", "float abs() const", asMETHOD(Complex,length), asCALL_THISCALL); assert( r >= 0 );
+
+	// Register the swizzle operators
+	r = engine->RegisterObjectMethod("complex", "complex get_ri() const", asMETHOD(Complex, get_ri), asCALL_THISCALL); assert( r >= 0 );
+	r = engine->RegisterObjectMethod("complex", "complex get_ir() const", asMETHOD(Complex, get_ir), asCALL_THISCALL); assert( r >= 0 );
+	r = engine->RegisterObjectMethod("complex", "void set_ri(const complex &in)", asMETHOD(Complex, set_ri), asCALL_THISCALL); assert( r >= 0 );
+	r = engine->RegisterObjectMethod("complex", "void set_ir(const complex &in)", asMETHOD(Complex, set_ir), asCALL_THISCALL); assert( r >= 0 );
+}
+
+void RegisterScriptMathComplex(asIScriptEngine *engine)
+{
+	if( strstr(asGetLibraryOptions(), "AS_MAX_PORTABILITY") )
+	{
+		assert( false );
+		// TODO: implement support for generic calling convention
+		// RegisterScriptMathComplex_Generic(engine);
+	}
+	else
+		RegisterScriptMathComplex_Native(engine);
+}
+
+END_AS_NAMESPACE
+
+

--- a/source/angelscript_addons/scriptmath/scriptmathcomplex.h
+++ b/source/angelscript_addons/scriptmath/scriptmathcomplex.h
@@ -1,0 +1,61 @@
+#ifndef SCRIPTMATHCOMPLEX_H
+#define SCRIPTMATHCOMPLEX_H
+
+#ifndef ANGELSCRIPT_H 
+// Avoid having to inform include path if header is already include before
+#include <angelscript.h>
+#endif
+
+
+BEGIN_AS_NAMESPACE
+
+// This class implements complex numbers and the common 
+// operations that can be done with it.
+//
+// Ref: http://mathworld.wolfram.com/ComplexNumber.html
+
+struct Complex
+{
+	Complex();
+	Complex(const Complex &other);
+	Complex(float r, float i = 0);
+
+	// Assignment operator
+	Complex &operator=(const Complex &other);
+	
+	// Compound assigment operators
+	Complex &operator+=(const Complex &other);
+	Complex &operator-=(const Complex &other);
+	Complex &operator*=(const Complex &other);
+	Complex &operator/=(const Complex &other);
+
+	float length() const;
+	float squaredLength() const;
+
+	// Swizzle operators
+	Complex get_ri() const;
+	void    set_ri(const Complex &in);
+	Complex get_ir() const;
+	void    set_ir(const Complex &in);
+
+	// Comparison
+	bool operator==(const Complex &other) const;
+	bool operator!=(const Complex &other) const;
+	
+	// Math operators
+	Complex operator+(const Complex &other) const;
+	Complex operator-(const Complex &other) const;
+	Complex operator*(const Complex &other) const;
+	Complex operator/(const Complex &other) const;
+
+	float r;
+	float i;
+};
+
+// This function will determine the configuration of the engine
+// and use one of the two functions below to register the string type
+void RegisterScriptMathComplex(asIScriptEngine *engine);
+
+END_AS_NAMESPACE
+
+#endif

--- a/source/angelscript_addons/scriptstdstring/scriptstdstring.cpp
+++ b/source/angelscript_addons/scriptstdstring/scriptstdstring.cpp
@@ -14,123 +14,80 @@ using namespace std;
 // Usually where the variables are only used in debug mode.
 #define UNUSED_VAR(x) (void)(x)
 
-#if AS_USE_STRINGPOOL == 1
-
 #ifdef AS_CAN_USE_CPP11
-	// The string pool doesn't need to keep a specific order in the
-	// pool, so the unordered_map is faster than the ordinary map
-	#include <unordered_map>  // std::unordered_map
+// The string factory doesn't need to keep a specific order in the
+// cache, so the unordered_map is faster than the ordinary map
+#include <unordered_map>  // std::unordered_map
 BEGIN_AS_NAMESPACE
-	typedef unordered_map<const char *, string> map_t;
+typedef unordered_map<string, int> map_t;
 END_AS_NAMESPACE
 #else
-	#include <map>      // std::map
+#include <map>      // std::map
 BEGIN_AS_NAMESPACE
-	typedef map<const char *, string> map_t;
+typedef map<string, int> map_t;
 END_AS_NAMESPACE
 #endif
 
 BEGIN_AS_NAMESPACE
 
-// By keeping the literal strings in a pool the application
-// performance is improved as there are less string copies created.
-
-// The string pool will be kept as user data in the engine. We'll
-// need a specific type to identify the string pool user data.
-// We just define a number here that we assume nobody else is using for
-// object type user data. The add-ons have reserved the numbers 1000
-// through 1999 for this purpose, so we should be fine.
-const asPWORD STRING_POOL = 1001;
-
-// This global static variable is placed here rather than locally within the
-// StringFactory, due to memory leak detectors that don't see the deallocation
-// of global variables. By placing the variable globally it will be initialized
-// before the memory leak detector starts, thus it won't report the missing
-// deallocation. An example of this the Marmalade leak detector initialized with
-// IwGxInit() and finished with IwGxTerminate().
-static const string emptyString;
-
-static const string &StringFactory(asUINT length, const char *s)
+class CStdStringFactory : public asIStringFactory
 {
-	// Each engine instance has its own string pool
-	asIScriptContext *ctx = asGetActiveContext();
-	if( ctx == 0 )
+public:
+	CStdStringFactory() {}
+	~CStdStringFactory() 
 	{
-		// The string factory can only be called from a script
-		assert( ctx );
-		return emptyString;
-	}
-	asIScriptEngine *engine = ctx->GetEngine();
-
-	map_t *pool = reinterpret_cast< map_t* >(engine->GetUserData(STRING_POOL));
-	if( !pool )
-	{
-		// The string pool hasn't been created yet, so we'll create it now
-		asAcquireExclusiveLock();
-
-		// Make sure the string pool wasn't created while we were waiting for the lock
-		pool = reinterpret_cast< map_t* >(engine->GetUserData(STRING_POOL));
-		if( !pool )
-		{
-			#if defined(__S3E__)
-			pool = new map_t;
-			#else
-			pool = new (nothrow) map_t;
-			#endif
-			if( pool == 0 )
-			{
-				ctx->SetException("Out of memory");
-				asReleaseExclusiveLock();
-				return emptyString;
-			}
-			engine->SetUserData(pool, STRING_POOL);
-		}
-
-		asReleaseExclusiveLock();
+		// The script engine must release each string 
+		// constant that it has requested
+		assert(stringCache.size() == 0);
 	}
 
-	// We can't let other threads modify the pool while we query it
-	asAcquireSharedLock();
-
-	// First check if a string object hasn't been created already
-	map_t::iterator it;
-	it = pool->find(s);
-	if( it != pool->end() )
+	const void *GetStringConstant(const char *data, asUINT length)
 	{
-		asReleaseSharedLock();
-		return it->second;
+		string str(data, length);
+		map_t::iterator it = stringCache.find(str);
+		if (it != stringCache.end())
+			it->second++;
+		else
+			it = stringCache.insert(map_t::value_type(str, 1)).first;
+
+		return reinterpret_cast<const void*>(&it->first);
 	}
 
-	asReleaseSharedLock();
-
-	// Acquire an exclusive lock so we can add the new string to the pool
-	asAcquireExclusiveLock();
-
-	// Make sure the string wasn't created while we were waiting for the exclusive lock
-	it = pool->find(s);
-	if( it == pool->end() )
+	int  ReleaseStringConstant(const void *str)
 	{
-		// Create a new string object
-		it = pool->insert(map_t::value_type(s, string(s, length))).first;
+		if (str == 0)
+			return asERROR;
+
+		map_t::iterator it = stringCache.find(*reinterpret_cast<const string*>(str));
+		if (it == stringCache.end())
+			return asERROR;
+
+		it->second--;
+		if (it->second == 0)
+			stringCache.erase(it);
+		return asSUCCESS;
 	}
 
-	asReleaseExclusiveLock();
-	return it->second;
-}
+	int  GetRawStringData(const void *str, char *data, asUINT *length) const
+	{
+		if (str == 0)
+			return asERROR;
 
-static void CleanupEngineStringPool(asIScriptEngine *engine)
-{
-	map_t *pool = reinterpret_cast< map_t* >(engine->GetUserData(STRING_POOL));
-	if( pool )
-		delete pool;
-}
+		if (length)
+			*length = (asUINT)reinterpret_cast<const string*>(str)->length();
 
-#else
-static string StringFactory(asUINT length, const char *s)
-{
-	return string(s, length);
-}
-#endif
+		if (data)
+			memcpy(data, reinterpret_cast<const string*>(str)->c_str(), reinterpret_cast<const string*>(str)->length());
+
+		return asSUCCESS;
+	}
+
+	// TODO: Make sure the access to the string cache is thread safe
+	map_t stringCache;
+};
+
+static CStdStringFactory stringFactory;
+
 
 static void ConstructString(string *thisPointer)
 {
@@ -739,16 +696,7 @@ void RegisterStdString_Native(asIScriptEngine *engine)
 	r = engine->RegisterObjectType("string", sizeof(string), asOBJ_VALUE | asOBJ_APP_CLASS_CDAK); assert( r >= 0 );
 #endif
 
-#if AS_USE_STRINGPOOL == 1
-	// Register the string factory
-	r = engine->RegisterStringFactory("const string &", asFUNCTION(StringFactory), asCALL_CDECL); assert( r >= 0 );
-
-	// Register the cleanup callback for the string pool
-	engine->SetEngineUserDataCleanupCallback(CleanupEngineStringPool, STRING_POOL);
-#else
-	// Register the string factory
-	r = engine->RegisterStringFactory("string", asFUNCTION(StringFactory), asCALL_CDECL); assert( r >= 0 );
-#endif
+	r = engine->RegisterStringFactory("string", &stringFactory);
 
 	// Register the object operator overloads
 	r = engine->RegisterObjectBehaviour("string", asBEHAVE_CONSTRUCT,  "void f()",                    asFUNCTION(ConstructString), asCALL_CDECL_OBJLAST); assert( r >= 0 );
@@ -839,26 +787,6 @@ void RegisterStdString_Native(asIScriptEngine *engine)
 	// replaceRange - replaces a range of bytes in the string
 	// multiply/times/opMul/opMul_r - takes the string and multiplies it n times, e.g. "-".multiply(5) returns "-----"
 }
-
-#if AS_USE_STRINGPOOL == 1
-static void StringFactoryGeneric(asIScriptGeneric *gen)
-{
-	asUINT length = gen->GetArgDWord(0);
-	const char *s = (const char*)gen->GetArgAddress(1);
-
-	// Return a reference to a string
-	gen->SetReturnAddress(const_cast<string*>(&StringFactory(length, s)));
-}
-#else
-static void StringFactoryGeneric(asIScriptGeneric *gen)
-{
-	asUINT length = gen->GetArgDWord(0);
-	const char *s = (const char*)gen->GetArgAddress(1);
-
-	// Return a string value
-	new (gen->GetAddressOfReturnLocation()) string(StringFactory(length, s));
-}
-#endif
 
 static void ConstructStringGeneric(asIScriptGeneric * gen)
 {
@@ -1288,16 +1216,7 @@ void RegisterStdString_Generic(asIScriptEngine *engine)
 	// Register the string type
 	r = engine->RegisterObjectType("string", sizeof(string), asOBJ_VALUE | asOBJ_APP_CLASS_CDAK); assert( r >= 0 );
 
-#if AS_USE_STRINGPOOL == 1
-	// Register the string factory
-	r = engine->RegisterStringFactory("const string &", asFUNCTION(StringFactoryGeneric), asCALL_GENERIC); assert( r >= 0 );
-
-	// Register the cleanup callback for the string pool
-	engine->SetEngineUserDataCleanupCallback(CleanupEngineStringPool, STRING_POOL);
-#else
-	// Register the string factory
-	r = engine->RegisterStringFactory("string", asFUNCTION(StringFactoryGeneric), asCALL_GENERIC); assert( r >= 0 );
-#endif
+	r = engine->RegisterStringFactory("string", &stringFactory);
 
 	// Register the object operator overloads
 	r = engine->RegisterObjectBehaviour("string", asBEHAVE_CONSTRUCT,  "void f()",                    asFUNCTION(ConstructStringGeneric), asCALL_GENERIC); assert( r >= 0 );

--- a/source/angelscript_addons/scriptstdstring/scriptstdstring.h
+++ b/source/angelscript_addons/scriptstdstring/scriptstdstring.h
@@ -22,16 +22,6 @@
 // Compilation settings
 //
 
-// The use of the string pool can improve performance quite drastically
-// for scripts that work with a lot of literal string constants. 
-//
-//  1 = on
-//  0 = off
-
-#ifndef AS_USE_STRINGPOOL
-#define AS_USE_STRINGPOOL 1
-#endif
-
 // Sometimes it may be desired to use the same method names as used by C++ STL.
 // This may for example reduce time when converting code from script to C++ or
 // back.

--- a/source/main/scripting/CBytecodeStream.cpp
+++ b/source/main/scripting/CBytecodeStream.cpp
@@ -35,18 +35,20 @@ CBytecodeStream::~CBytecodeStream()
         fclose(f);
 }
 
-void CBytecodeStream::Write(const void* ptr, AngelScript::asUINT size)
+int CBytecodeStream::Write(const void* ptr, AngelScript::asUINT size)
 {
     if (!f)
-        return;
+        return -1;
     size_t result = fwrite(ptr, size, 1, f);
+    return 0;
 }
 
-void CBytecodeStream::Read(void* ptr, AngelScript::asUINT size)
+int CBytecodeStream::Read(void* ptr, AngelScript::asUINT size)
 {
     if (!f)
-        return;
+        return -1;
     size_t result = fread(ptr, size, 1, f);
+    return 0;
 }
 
 bool CBytecodeStream::Existing()

--- a/source/main/scripting/CBytecodeStream.h
+++ b/source/main/scripting/CBytecodeStream.h
@@ -34,8 +34,8 @@ class CBytecodeStream : public AngelScript::asIBinaryStream
 public:
     CBytecodeStream(std::string filename);
     ~CBytecodeStream();
-    void Read(void* ptr, AngelScript::asUINT size);
-    void Write(const void* ptr, AngelScript::asUINT size);
+    int Read(void* ptr, AngelScript::asUINT size);
+    int Write(const void* ptr, AngelScript::asUINT size);
     bool Existing();
 private:
     FILE* f;

